puzzles/views.py

```diff
@@ -190,24 +197,25 @@ def delete_puzzle(request, pk):
 
 @require_POST
 @login_required(login_url='/accounts/login/')
+@transaction.atomic
 def add_tag(request, pk):
     form = TagForm(request.POST)
     if not form.is_valid():
         messages.error(request, form)
         return HttpResponseRedirect(request.META.get('HTTP_REFERER', '/'))
-    puzzle = get_object_or_404(Puzzle, pk=pk)
+    puzzle = get_object_or_404(Puzzle.objects.select_for_update(), pk=pk)
     (tag, _) = PuzzleTag.objects.update_or_create(
         name=form.cleaned_data["name"],
         defaults={'color' : form.cleaned_data["color"]}
     )
     if tag.is_meta:
-        metapuzzle = Puzzle.objects.get(name=tag.name)
+        metapuzzle = get_object_or_404(Puzzle, name=tag.name)
```

Comment:
Do we need to use `Puzzle.objects.select_for_update()` here as well?

Reply:
Hmm, I'm not sure. I didn't think we needed it since we aren't updating metapuzzle.

Reply:
Could an inconsistency arise if the metapuzzle were changed to a regular puzzle while this transaction was happening? Then on line 218, we would be treating the now-regular puzzle as the meta for this puzzle?

Reply:
Yeah you're right, a change to metapuzzle could happen from a third puzzle.